internal/libs/reservoir/reservoir.go

```diff
@@ -50,7 +57,9 @@ func (s *Sampler[T]) Add(item T) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	s.dirtySinceAdd = true
```

Comment:
nit: you can condition the dirty bit on whether the sample set has actually changed

Reply:
I mean, you set dirtySinceAdd iff one of the following "ifs" evaluates to true (either append or replace actually happens)